app/javascript/packages/document-capture/components/full-screen.jsx

```diff
@@ -1,81 +1,139 @@
-import { useRef, useEffect, useCallback } from 'react';
+import { useRef, useEffect } from 'react';
+import { createPortal } from 'react-dom';
 import { createFocusTrap } from 'focus-trap';
 import useI18n from '../hooks/use-i18n';
 import useAsset from '../hooks/use-asset';
 
+/** @typedef {import('focus-trap').FocusTrap} FocusTrap */
 /** @typedef {import('react').ReactNode} ReactNode */
 
 /**
- * @typedef FullScreenProps
- *
- * @prop {()=>void=} onRequestClose Callback invoked when user initiates close intent.
- * @prop {ReactNode} children       Child elements.
+ * @typedef {()=>void} RequestCloseCallback
  */
 
 /**
- * Number of active instances of FullScreen currently mounted, used in determining when overlay body
- * class should be added or removed.
+ * @typedef FullScreenProps
  *
- * @type {number}
+ * @prop {RequestCloseCallback=} onRequestClose Callback invoked when user initiates close intent.
+ * @prop {string} label Accessible label for modal.
+ * @prop {ReactNode} children Child elements.
  */
-let activeInstances = 0;
 
 /**
- * @param {FullScreenProps} props Props object.
+ * @param {React.MutableRefObject<HTMLElement?>} containerRef
+ * @param {RequestCloseCallback} onRequestClose
  */
-function FullScreen({ onRequestClose = () => {}, children }) {
-  const { t } = useI18n();
-  const { getAssetPath } = useAsset();
-  const trapRef = useRef(/** @type {?import('focus-trap').FocusTrap} */ (null));
+function useFocusTrap(containerRef, onRequestClose) {
+  const trapRef = useRef(/** @type {FocusTrap?} */ (null));
   const onRequestCloseRef = useRef(onRequestClose);
+
   useEffect(() => {
     // Since the focus trap is only initialized once, but the callback could
     // be changed, ensure that the current reference is kept as a mutable value
     // to reference in the deactivation.
     onRequestCloseRef.current = onRequestClose;
   }, [onRequestClose]);
 
-  const setFocusTrapRef = useCallback((node) => {
-    if (trapRef.current) {
-      trapRef.current.deactivate();
-    }
-
-    if (node) {
-      trapRef.current = createFocusTrap(node, {
+  useEffect(() => {
+    if (containerRef.current) {
+      trapRef.current = createFocusTrap(containerRef.current, {
         onDeactivate: () => onRequestCloseRef.current(),
         clickOutsideDeactivates: true,
       });
 
       trapRef.current.activate();
     }
-  }, []);
-
-  useEffect(() => {
-    if (activeInstances++ === 0) {
-      document.body.classList.add('has-full-screen-overlay');
-      document.documentElement.classList.add('has-full-screen-overlay');
-    }
 
     return () => {
-      if (--activeInstances === 0) {
-        document.body.classList.remove('has-full-screen-overlay');
-        document.documentElement.classList.remove('has-full-screen-overlay');
-      }
+      trapRef.current?.deactivate();
     };
   }, []);
+}
+
+const useBodyClass = (() => {
```

Comment:
Note: I tried avoiding the weird IIFE / closure here with a property on the function itself:

```js
function useBodyClass(className, Component) {
  const activeInstancesByComponent = useBodyClass.activeInstancesByComponent ??= new WeakMap;

  // ...
}
```

...but I find that, with our JSDoc-based usage of TypeScript, it's quite difficult to define custom properties for functions.

The closest I got was casting it with a type intersection:

```js
/** @type {typeof useBodyClass & {activeInstancesByCount: WeakMap<*, number>}} */ (useBodyClass)
```

At which point I struggled with:

- Needing to cast every reference of `useBodyClass`, or assign a separate `useEnhancedBodyClass` variable just to retain the type.
- Prettier formatting doing the wrong thing and breaking the TypeScript casting

Reply:
> Note: I tried avoiding the weird IIFE / closure here with a property on the function itself:
>
>...but I find that, with our JSDoc-based usage of TypeScript, it's quite difficult to define custom properties for functions.

Seems to work fine if we assign it at the top-level 🤔 

```js
function useBodyClass(className, Component) {
  // ...
}

useBodyClass.activeInstancesByComponent = /** @type {WeakMap<*, number>} */ (new WeakMap());
```

Reply:
nothing is stopping the map from being a top-level declaration in the file right? While it would technically be accessible to other functions in this file, that could be Ok? Or we could just move it to its own file and not export it right?

FWIW I didn't mind the IIFE, but if we do keep it, maybe could copy the `@param... @return` JSDOC to the outside of it rather than having to look inside a and down a few lines to see what it returns?

Reply:
> nothing is stopping the map from being a top-level declaration in the file right? While it would technically be accessible to other functions in this file, that could be Ok? Or we could just move it to its own file and not export it right?

Yep, in the context of this file, it can be messy in how it's a shared context. Less messy once we extract this to its own file. Done in 44029b1db.